tests/unittests/GraphOptzTest.cpp

```diff
@@ -659,54 +689,69 @@ TEST_F(GraphOptz, sinkTransposeBelowRescale) {
   EXPECT_EQ(convTRInput.dims(), llvm::makeArrayRef(origDims));
   EXPECT_EQ(convTRInput.getNode()->getNthInput(0).dims(),
             llvm::makeArrayRef(origDims));
-  EXPECT_EQ(F_->getNodes().size(), 3);
+  EXPECT_EQ(originalF_->getNodes().size(), 3);
 }
 
 TEST_F(GraphOptz, cancelTwoTransposes) {
   const size_t origDims[] = {1, 5, 10, 15};
-  Node *A = mod_.createPlaceholder(ElemKind::FloatTy, origDims, "input", false);
-  Node *T1 = F_->createTranspose("transpose", A, NCHW2NHWC);
-  Node *T2 = F_->createTranspose("transpose", T1, NHWC2NCHW);
-  ReluNode *K = F_->createRELU("relu", T2);
-  SaveNode *save = F_->createSave("ret", K);
+  Placeholder *A =
+      mod_->createPlaceholder(ElemKind::FloatTy, origDims, "input", false);
+  Node *T1 = originalF_->createTranspose("transpose", A, NCHW2NHWC);
+  Node *T2 = originalF_->createTranspose("transpose", T1, NHWC2NCHW);
+  ReluNode *K = originalF_->createRELU("relu", T2);
+  SaveNode *save = originalF_->createSave("ret", K);
 
   EXPECT_EQ(K->getInput().dims(), llvm::makeArrayRef(origDims));
-  EXPECT_EQ(F_->getNodes().size(), 4);
+  EXPECT_EQ(originalF_->getNodes().size(), 4);
+
+  optimizedF_ = originalF_->clone(originalF_->getName().str() + "_optmized");
+  ::glow::optimize(optimizedF_, CompilationMode::Infer);
 
-  ::glow::optimize(F_, CompilationMode::Infer);
+  EXPECT_EQ(optimizedF_->getNodes().size(), 2);
+
+  for (auto &N : optimizedF_->getNodes()) {
+    if (N.getKind() == Kinded::Kind::SaveNodeKind) {
+      save = llvm::dyn_cast<SaveNode>(&N);
+    }
+  }
 
-  EXPECT_EQ(F_->getNodes().size(), 2);
   ReluNode *relu = llvm::dyn_cast<ReluNode>(save->getInput());
   ASSERT_TRUE(relu);
   EXPECT_EQ(relu->getResult().dims(), llvm::makeArrayRef(origDims));
   EXPECT_EQ(relu->getInput().getNode(), A);
+
+  bindings_.allocate(mod_->getPlaceholders());
```

Comment:
Are these two lines going to be test-specific? If you mean here something generic like: "Allocate all placeholders and initialize them with random values" and then check for the numerical equivalence, then this logic could be moved into a test-independent helper function. 

BTW, using `randomize` makes this test non-deterministic, or? I.e. if we run it multiple times, it may eventually fail or not fail depending on the values of the placeholders. So, should we make it more deterministic by seeding the RNG in a special way? Also, may be we need to run the tests not only on a single combination of random placeholder values, but on a set of such combinations?

Reply:
> Are these two lines going to be test-specific? If you mean here something generic like: "Allocate all placeholders and initialize them with random values" and then check for the numerical equivalence, then this logic could be moved into a test-independent helper function.

I think some tests (such as this one) do not care about the specific values and can use random values. But if the optimisation is something like replacing `MaxPool -> Relu` with `Relu -> MaxPool`, I think the test writer may want to make sure that it is possible to specify that some of the input values should be negative to test the clamping behaviour of the relu operation.

> BTW, using randomize makes this test non-deterministic, or? I.e. if we run it multiple times, it may eventually fail or not fail depending on the values of the placeholders. So, should we make it more deterministic by seeding the RNG in a special way?

The PRNG inside the module is already seeded with some default seed, so I don't think we have to worry about non-determinism:
```
                     llvm::cl::init(glow::PseudoRNG::Engine::default_seed));
```

> Also, may be we need to run the tests not only on a single combination of random placeholder values, but on a set of such combinations?

What did you have in mind?

Reply:
>> Also, may be we need to run the tests not only on a single combination of random placeholder values, but on a set of such combinations?

> What did you have in mind?

Like literally have a loop around `randomize inputs; checkNumericalEquivalence;` so that we check the numerical equivalence for a multiple rounds of random inputs. Would it improve our confidence in the equivalence of results?

Reply:
I think it would, but we should prefer deliberate specification of input whenever possible. Consider the `MaxPool -> Relu` ==> `Relu -> MaxPool` mentioned earlier. If the input values were randomly generated and all came out positive (either due to bad luck or careless specification of the upper and lower bound of the RNG) and the optimization code was incorrect and omitted the Relu, a numerical equivalence check would pass.